pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/PandioBandwidthHandler.java

```diff
@@ -0,0 +1,187 @@
+package org.apache.pulsar.proxy.server;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import lombok.Builder;
+import lombok.Data;
+import lombok.With;
+import lombok.val;
+import org.apache.pulsar.common.api.proto.PulsarApi;
+import org.apache.pulsar.common.naming.TopicName;
+import org.apache.pulsar.common.util.protobuf.ByteBufCodedInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.*;
+import java.util.Map;
+
+public class PandioBandwidthHandler extends ChannelInboundHandlerAdapter {
+
+    public static final String HANDLER_NAME = "pandioBandwidthPublisher";
+    //producerid+channelid as key
+    //or consumerid+channelid as key
+
+    private static final Map<String, String> producerHashMap = new ConcurrentHashMap<>();
+    private static final Map<String, String> consumerHashMap = new ConcurrentHashMap<>();
+
+    private static final Map<String, List<ChannelInfo>> channelMap = new ConcurrentHashMap<>();
+
+    public static Map<String, String> inboundOutboundChannelMap = new ConcurrentHashMap<>();
+
+    private static ExecutorService executorService = null;
+    private final ProxyConfiguration config;
+
+    public PandioBandwidthHandler(final ProxyConfiguration config) {
+        this.config = config;
+        if (PandioBandwidthHandler.executorService == null) {
+            PandioBandwidthHandler.executorService = Executors.newFixedThreadPool(config.getPandioBandwidthPublisherNumOfThreads());
+        }
+    }
+
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        ByteBuf buffer = (ByteBuf) (msg);
```

Comment:
are we always sure that `ByteBuf` is the type passed? should we wrap this in try catch?

Reply:
No need. It is bytebuf. They are passing object because they are overriding an interface which doesn't use Generics so Object everywhere.